src/toolkit/res_map.h

```diff
@@ -0,0 +1,189 @@
+#ifndef CYCLUS_SRC_TOOLKIT_RES_MAP_H_
+#define CYCLUS_SRC_TOOLKIT_RES_MAP_H_
+
+#include <iomanip>
+#include <limits>
+#include <map>
+#include <set>
+
+#include "cyc_arithmetic.h"
+#include "cyc_limits.h"
+#include "error.h"
+#include "product.h"
+#include "material.h"
+#include "resource.h"
+#include "res_manip.h"
+
+namespace cyclus {
+namespace toolkit {
+
+/// ResMap     is a helper class that provides semi-automated management of
+/// a collection of resources (e.g. agent stocks and inventories).
+/// Constructed buffers have infinite capacity unless explicitly changed.
+/// Resource popping occurs in the order the resources were pushed (i.e. oldest
+/// resources are popped first), unless explicitly specified otherwise.
+///
+/// Typically, a ResMap     will be a member variable on an agent/archetype class.
+/// Resources can be added and retrieved from it as needed, and the buffer can
+/// be queried in various ways as done in the example below:
+///
+/// @code
+/// class MyAgent : public cyclus::Facility {
+///  public:
+///   Tick() {
+///     double batch_size = 2703;
+///     if (outventory_.space() < batch_size) {
+///       return;
+///     } else if (inventory_.quantity() < batch_size) {
+///       return;
+///     }
+///
+///     outventory_.Push(inventory_.Pop(batch_size));
+///   }
+///
+///   ... // resource exchange to fill up inventory_ buffer
+///
+///  private:
+///   ...
+///   cyclus::toolkit::ResMap    <cyclus::Material> inventory_;
+///   cyclus::toolkit::ResMap    <cyclus::Material> outventory_;
+/// };
+/// @endcode
+///
+/// In this example, if there is sufficient material in inventory_, 2703 kg is
+/// removed as a single object that is then placed in another buffer
+/// (outventory_) each time step.
```

Comment:
I think the comment needs to effectively be rewritten for ResMap, but @scopatz pointed that out in his original comment.